[3007.x] Add started event to PublishServer

dwoz opened the pull request:
- Add a started event that gets set after the server transport is ready for clients to connect.
- Wait on publish servers start while the master process is starting up.

### What does this PR do?

### What issues does this PR fix or reference?
Fixes #66993

### Previous Behavior

Publish clients in the master process could encounter exceptions while trying to connect.

### New Behavior

Publish clients within the master processes connect without error.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

dwoz committed:
Add started event to PublishServer

- Add a started event that gets set after the server transport is ready
for clients to connect.
- Wait on publish servers start while the master process is starting up.

dwoz committed:
Remove un-needed param

dwoz committed:
Increase timeout for windows tests

Sxderp commented:
1. I'm not sure the rational for passing this as an argument. What's the use case for using an arbitrary object? I think it would be better to have a separate method that you can give a timeout to so that transports can implement their own methodology and better integrate with #66751. Something like `def wait_startup(timeout): bool`
2. I think #66931 is needed for the tests to fully pass.

dwoz committed:
Pull in changes from #66931

Make sure we only call chmod when pull_path or pub_path is used. Add
regression tests for this behavior.

dwoz committed:
Improve tests

dwoz committed:
Skip tests that are not applicable on windows

dwoz commented:
> 1. I'm not sure the rational for passing this as an argument. What's the use case for using an arbitrary object? I think it would be better to have a separate method that you can give a timeout to so that transports can implement their own methodology and better integrate with [[3007.x] transports: fix and normalize PublishServer method signatures #66751](https://github.com/saltstack/salt/pull/66751). Something like `def wait_startup(self, timeout): bool`
> 2. I think [[3007.x] transport: tcp check that pull_path exists before chmod #66931](https://github.com/saltstack/salt/pull/66931) is needed for the tests to fully pass.

I agree a method like `wait_startup` might make sense in the future.
